source/shadow-tasks/shadow_device_task.c

```diff
@@ -0,0 +1,801 @@
+ * 2. Subscribe to those MQTT topics using the MQTT Agent.
+ * 3. Register callbacks for incoming shadow topic publishes with the subsciption_manager.
+ * 3. Publish to report the current state of powerOn.
+ * 5. Check if powerOn has been changed and send an update if so.
+ * 6. If a publish to update reported state was sent, wait until either prvIncomingPublishUpdateAcceptedCallback
+ *    or prvIncomingPublishUpdateRejectedCallback handle the response.
+ * 7. Wait until time for next check and repeat from step 5.
+ *
+ * Meanwhile, when prvIncomingPublishUpdateDeltaCallback recives changes to shadow state, it will apply them on the
+ * device
+ */
+
+/* Standard includes. */
+#include <stdlib.h>
+#include <string.h>
+#include <stdio.h>
+
+/* Kernel includes. */
+#include "FreeRTOS.h"
+#include "task.h"
+
+/* Demo config. */
+#include "demo_config.h"
+
+/* MQTT library includes. */
+#include "mqtt_agent.h"
+
+/* Subscription manager header include. */
+#include "subscription_manager.h"
+
+/* JSON library includes. */
+#include "core_json.h"
+
+/* Shadow API header. */
+#include "shadow.h"
+
+/**
+ * democonfigCLIENT_IDENTIFIER is required. Throw compilation error if it is not defined.
+ */
+#ifndef democonfigCLIENT_IDENTIFIER
+    #error "Please define democonfigCLIENT_IDENTIFIER in demo_config.h to the thing name registered with AWS IoT Core."
+#endif
+
+/**
+ * @brief Format string representing a Shadow document with a "reported" state.
+ *
+ * The real json document will look like this:
+ * {
+ *   "state": {
+ *     "reported": {
+ *       "powerOn": 1
+ *     }
+ *   },
+ *   "clientToken": "021909"
+ * }
+ *
+ * Note the client token, which is required for all Shadow updates. The client
+ * token must be unique at any given time, but may be reused once the update is
+ * completed. For this demo, a timestamp is used for a client token.
+ */
+#define shadowexampleSHADOW_REPORTED_JSON \
+    "{"                                   \
+    "\"state\":{"                         \
+    "\"reported\":{"                      \
+    "\"powerOn\":%01u"                    \
+    "}"                                   \
+    "},"                                  \
+    "\"clientToken\":\"%06lu\""           \
+    "}"
+
+/**
+ * @brief The expected size of #SHADOW_REPORTED_JSON.
+ *
+ * Because all the format specifiers in #SHADOW_REPORTED_JSON include a length,
+ * its full size is known at compile-time by pre-calculation.
+ *
+ * Because all the format specifiers in #SHADOW_REPORTED_JSON include a length,
+ * its full actual size is known by pre-calculation, here's the formula why
+ * the length need to minus 3:
+ * 1. The length of "%01d" is 4.
+ * 2. The length of %06lu is 5.
+ * 3. The actual length we will use in case 1. is 1 ( for the state of powerOn ).
+ * 4. The actual length we will use in case 2. is 6 ( for the clientToken length ).
+ * 5. Thus the additional size 3 = 4 + 5 - 1 - 6 + 1 (termination character).
```

Comment:
```suggestion
 * Since all of the format specifiers in #SHADOW_REPORTED_JSON include a length,
 * its actual size can be pre-calculated at compile time from the difference between
 * the lengths of the format strings and their formatted output. We must subtract 3
 * from the length as according the following formula:
 * 1. The length of the format string "%01u" is 4.
 * 2. The length of the format string "%06lu" is 5.
 * 3. The formatted length in case 1. is 1 ( for the state of powerOn ).
 * 4. The formatted length in case 2. is 6 ( for the clientToken length ).
 * 5. Thus the additional size of our format is 3 = 4 + 5 - 1 - 6 + 1 (termination character).
```

Reply:
This wasn't changed